[LTC] Code-gen native_layer_norm

alanwaketan opened the pull request:
Summary:
This pull request code-generated native_layer_norm. In order to do that, it also:
1) introduced fixes for optional Tensor inputs,
2) mandated parameters being populated in the same order defined in the native function schema,
3) moved at::dtype and at::shape from leaf IR nodes to the base Node to avoid storing those two common fields in every leaf IR node's definitions.
4) introduced tuple output support, and code-gen `topk` to validate the implementation.
5) introduced optional operand support, and code-gen `native_layer_norm` to validate the implementation.
6) code-gen `native_layer_norm_backward`.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestLayerNorm*:AtenLtcTsTensorTest.TestTopK

Fixes #65837.

Commit:
[LTC] Move at::dtype and at::shape from leaf IR nodes to the base Node

Summary:
This commit move at::dtype and at::shape from leaf IR nodes to the base Node such that we avoid storing those two common fields in every leaf IR node's definitions.

Test Plan:
[.../lazy_tensor_core] test/cpp/build/test_ptltc

Commit:
[LTC] Code-gen layer_norm

Summary:
This pull request code-generated layer_norm. In order to do that, it also i) introduced fixes for optional Tensor inputs, and ii) mandated parameters being populated in the same order defined in the native function schema.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestLayerNorm

wconstab reviewed (approved):
this looks great, thanks @alanwaketan!!!

Commit:
[LTC] Supported optional operands in the code-gen and generate native_layer_norm

Summary:
This commit added support for optional operands in the code-gen engine, which basically introduced a named bitmap in each leaf node class to indicate whether the operand exists or not.
So the Clone() method and the lowering code can iterate over operands accordingly.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestLayerNorm

Commit:
[LTC] Code-gen native_layer_norm_backward

Summary:
This commit code-generated native_layer_norm_backward, and improved the code-gen engine a bit on how to convert VectorCType to std::vector<>.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestLayerNormBackward

alanwaketan commented:
@wconstab I have pushed two commits after your last approval. Don't forget to review them.😃

Commit:
[LTC] Add support for code-gen operators that have tuple output

Summary:
This patch added support for code-gen operators that have tuple output.
1) It changed Node::at_dtypes_ & Node::at_shapes_ to std::vector so now they are generalized to describe tuple output.
2) It added utilities to convert tuple to vector and vice versa.
3) It teached the code-gen engine to tell such an operator and utilize the above utilities.
4) It then generated topk.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestTopK
